File: server-dotnet/PokerPlanning.Api/Services/OAuthStateStore.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;

namespace PokerPlanning.Api.Services;

public sealed class OAuthStartState
{
    public required string FirebaseUid { get; init; }
    public required string ReturnUrl { get; init; }
    public required long CreatedAtMs { get; init; }
}

/// <summary>In-memory OAuth state (replace with Redis for multi-instance).</summary>
public sealed class OAuthStateStore
{
    private static readonly ConcurrentDictionary<string, OAuthStartState> Store = new();
    private const int TtlMs = 15 * 60 * 1000;

    private static void Sweep()
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        foreach (var kv in Store)
        {
            if (now - kv.Value.CreatedAtMs > TtlMs)
            {
                Store.TryRemove(kv.Key, out _);
            }
        }
    }

    public string Create(string firebaseUid, string returnUrl)
    {
        Sweep();
        var id = Convert.ToBase64String(RandomNumberGenerator.GetBytes(24))
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
        var state = new OAuthStartState
        {
            FirebaseUid = firebaseUid,
            ReturnUrl = returnUrl,
            CreatedAtMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        };
        Store[id] = state;
        return id;
    }

    public OAuthStartState? Consume(string id)
    {
        Sweep();
        if (!Store.TryRemove(id, out var v))
        {
            return null;
        }

        if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - v.CreatedAtMs > TtlMs)
        {
            return null;
        }

        return v;
    }
}

File: server/src/lib/oauthStateStore.ts
import { randomBytes } from 'node:crypto';

export interface OAuthStartState {
  firebaseUid: string;
  returnUrl: string;
  createdAtMs: number;
}

/** In-memory OAuth state. Replace with Redis for multi-instance deployments. */
const store = new Map<string, OAuthStartState>();

const TTL_MS = 15 * 60 * 1000;

function sweep(): void {
  const now = Date.now();
  for (const [k, v] of store) {
    if (now - v.createdAtMs > TTL_MS) {
      store.delete(k);
    }
  }
}

export function createOAuthState(data: Omit<OAuthStartState, 'createdAtMs'>): string {
  sweep();
  const id = randomBytes(24).toString('base64url');
  store.set(id, { ...data, createdAtMs: Date.now() });
  return id;
}

export function consumeOAuthState(id: string): OAuthStartState | null {
  sweep();
  const v = store.get(id);
  if (!v) {
    return null;
  }
  if (Date.now() - v.createdAtMs > TTL_MS) {
    store.delete(id);
    return null;
  }
  store.delete(id);
  return v;
}
